src/engine/fileOpsIns.cpp

```diff
@@ -634,6 +637,200 @@ void DivEngine::loadSBI(SafeReader& reader, std::vector<DivInstrument*>& ret, St
     delete ins;
   }
 }
+
+void DivEngine::loadOPLI(SafeReader& reader, std::vector<DivInstrument*>& ret, String& stripPath) {
+  DivInstrument* ins = new DivInstrument;
+
+  try {
+    reader.seek(0, SEEK_SET);
+    String header = reader.readString(11);
+    if (header.compare("WOPL3-INST") == 0) {
+      uint16_t version = reader.readS();
+      if (version > 3) {
+        logW("Unknown OPLI version.");
+      }
+
+      reader.readC();  // skip isPerc field
+
+      ins->type = DIV_INS_OPL;
+      String insName = reader.readString(32);
+      insName = (insName.length() > 0) ? insName : stripPath;
+      ins->name = insName;
+      reader.seek(7, SEEK_CUR);  // skip MIDI params
+      uint8_t instTypeFlags = reader.readC();  // [0EEEDCBA] - see WOPL/OPLI spec
+
+      bool is_4op = ((instTypeFlags & 0x1) == 1);
+      bool is_2x2op = (((instTypeFlags>>1) & 0x1) == 1);
+      bool is_rhythm = (((instTypeFlags>>4) & 0x7) > 0);
+
+      auto readOpliOp = [](SafeReader& reader, DivInstrumentFM::Operator& op) {
+        uint8_t characteristics = reader.readC();
+        uint8_t keyScaleLevel = reader.readC();
+        uint8_t attackDecay = reader.readC();
+        uint8_t sustainRelease = reader.readC();
+        uint8_t waveSelect = reader.readC();
+
+        op.mult = characteristics & 0xF;
+        op.ksr = ((characteristics >> 4) & 0x1);
+        op.sus = ((characteristics >> 5) & 0x1);
+        op.vib = ((characteristics >> 6) & 0x1);
+        op.am = ((characteristics >> 7) & 0x1);
+        op.tl = keyScaleLevel & 0x3F;
+        op.ksl = ((keyScaleLevel >> 6) & 0x3);
+        op.ar = ((attackDecay >> 4) & 0xF);
+        op.dr = attackDecay & 0xF;
+        op.rr = sustainRelease & 0xF;
+        op.sl = ((sustainRelease >> 4) & 0xF);
+        op.ws = waveSelect;
+      };
+
+      uint8_t feedConnect = reader.readC();
+      uint8_t feedConnect2nd = reader.readC();
+
+      ins->fm.alg = (feedConnect & 0x1);
+      ins->fm.fb = ((feedConnect >> 1) & 0xF);
+
+      if (is_4op && !is_2x2op) {
+        ins->fm.ops = 4;
+        ins->fm.alg = (feedConnect & 0x1) | ((feedConnect2nd & 0x1) << 1);
+        for (int i : {2,0,3,1}) {
+          readOpliOp(reader, ins->fm.op[i]);
+        }
+      } else {
+        ins->fm.ops = 2;
+        for (int i : {1,0}) {
+          readOpliOp(reader, ins->fm.op[i]);
+        }
+        if (is_rhythm) {
+          ins->fm.opllPreset = (uint8_t)(1<<4);
+
+        } else if (is_2x2op) {
+          // Note: Pair detuning offset not mappable. Use E5xx effect :P
+          ins->name = fmt::format("{0} (1)", insName);
```

Comment:
Suggestion: use fmt::sprintf. Normal-style formatting isn't really my deal...

Reply:
You may be convinced otherwise in future if you need to reference the same variable in the same string, but OK changed.